storm-core/src/jvm/org/apache/storm/StormTimer.java

```diff
@@ -0,0 +1,274 @@
+ */
+
+public class StormTimer {
+    private static final Logger LOG = LoggerFactory.getLogger(StormTimer.class);
+
+    public interface TimerFunc {
+        public void run(Object o);
+    }
+
+    public static class QueueEntry {
+        public final Long endTimeMs;
+        public final TimerFunc afn;
+        public final String id;
+
+        public QueueEntry(Long endTimeMs, TimerFunc afn, String id) {
+            this.endTimeMs = endTimeMs;
+            this.afn = afn;
+            this.id = id;
+        }
+    }
+
+    public static class StormTimerTask extends Thread {
+
+        private PriorityBlockingQueue<QueueEntry> queue = new PriorityBlockingQueue<QueueEntry>(10, new Comparator() {
+            @Override
+            public int compare(Object o1, Object o2) {
+                return ((QueueEntry)o1).endTimeMs.intValue() - ((QueueEntry)o2).endTimeMs.intValue();
+            }
+        });
+
+        // boolean to indicate whether timer is active
+        private AtomicBoolean active = new AtomicBoolean(false);
+
+        // function to call when timer is killed
+        private TimerFunc onKill;
+
+        //random number generator
+        private Random random = new Random();
+
+        // synchronization for cancelling the timer
+        private Semaphore cancelNotifier = new Semaphore(0);
+
+        //used for synchronization
+        private Object lock = new Object();
+
+        @Override
+        public void run() {
+            while (this.active.get()) {
+                QueueEntry queueEntry = null;
+                try {
+                    synchronized (this.lock) {
+                        queueEntry = this.queue.peek();
+                    }
+                    if ((queueEntry != null) && (Time.currentTimeMillis() >= queueEntry.endTimeMs)) {
+                        // It is imperative to not run the function
+                        // inside the timer lock. Otherwise, it is
+                        // possible to deadlock if the fn deals with
+                        // other locks, like the submit lock.
+                        synchronized (this.lock) {
+                            this.queue.poll();
+                        }
+                        queueEntry.afn.run(null);
+                    } else if (queueEntry != null) {
+                        //  If any events are scheduled, sleep until
+                        // event generation. If any recurring events
+                        // are scheduled then we will always go
+                        // through this branch, sleeping only the
+                        // exact necessary amount of time. We give
+                        // an upper bound, e.g. 1000 millis, to the
+                        // sleeping time, to limit the response time
+                        // for detecting any new event within 1 secs.
+                        Time.sleep(Math.min(1000, (queueEntry.endTimeMs - Time.currentTimeMillis())));
```

Comment:
Could you file a follow on JIRA so we can figure out a way to do a notify with simulate time, so we don't need this 1 second polling.